go/chat/signencrypt/codec.go

```diff
@@ -565,6 +566,59 @@ func OpenWhole(sealed []byte, encKey SecretboxKey, verifyKey VerifyKey, signatur
 	return append(output, moreOutput...), nil
 }
 
+// include verification of associated data. see
+// https://en.wikipedia.org/wiki/Authenticated_encryption#Authenticated_encryption_with_associated_data_(AEAD)
+
+type AEADMessage struct {
+	V int               `codec:"v" json:"v"`
+	A [sha512.Size]byte `codec:"a" json:"a"`
+	M []byte            `codec:"m" json:"m"`
+}
+
+// SealWithAssociatedData is a wrapper around SealWhole which adds an associatedData object
+// (see AEAD ciphers) which must be message-packable into bytes. This exact object is required
+// to call OpenWithAssociatedData on the ciphertext.
+func SealWithAssociatedData(msg []byte, associatedData interface{}, encKey SecretboxKey, signKey SignKey, signaturePrefix kbcrypto.SignaturePrefix, nonce Nonce) (ret []byte, err error) {
+	adEncoded, err := msgpack.Encode(associatedData)
+	if err != nil {
+		return ret, err
+	}
+	adHash := sha512.Sum512(adEncoded)
+	aeadMsg := AEADMessage{V: 1, A: adHash, M: msg}
+	clearBytes, err := msgpack.Encode(aeadMsg)
+	if err != nil {
+		return ret, err
+	}
+	return SealWhole(clearBytes, encKey, signKey, signaturePrefix, nonce), nil
```

Comment:
One idea might be have the output be a struct that itself contains the associated data. So that you have a single object having both the associated data and the encrypted one.
This is probably good enough though, as having the data part of the ciphertext could interfere with streaming decryption.

Reply:
yeah - seems like a lateral move. i'd rather not do this.